teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,308 @@
+```
+
+##### Http Resolver
+Based on `http` resolver's [params](https://tekton.dev/docs/pipelines/http-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+<http-url>?http-username=<http-username>&http-password-secret=<http-password-secret>&http-password-secret-key=<http-password-secret-key>
+```
+
+Here, the url contains the `scheme` and `location`: `<scheme>://<locaiton>`:
+
+Example:
+```
+http://example.com/foo/task.yaml
+https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml
+```
+
+Using the above example, a user can do the following when referencing the Task:
+```yaml
+taskRef:
+  resolver: http
+  name: "https://raw.githubusercontent.com/tektoncd-catalog/git-clone/main/task/git-clone/git-clone.yaml?http-username=git&http-password-secret=git-secret&http-password-secret-key=git-token"
+```
+
+##### Hub Resolver
+Based on `hub` resolver's [params](https://tekton.dev/docs/pipelines/hub-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+hub://<type>/<catalog>/<kind>/<name>@<version>
+```
+**Note** that we use a scheme called `hub` which is only meaningful to the hub resolver.
+
+Example:
+
+```
+hub://artifact/tekton-catalog-tasks/task/git-clone@0.6
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: hub
+  name: "hub://artifact/tekton-catalog-tasks/task/git-clone@0.6"
+```
+
+##### Cluster Resolver
+Based on `cluster` resolver's [params](https://tekton.dev/docs/pipelines/cluster-resolver/#parameters), we define the following `resolver-reference` format: 
+
+```
+cluster://<kind>/<namespace>/<name>
+```
+
+**Note** that we use a scheme called `cluster` which is only meaningful to the hub resolver.
+
+Example:
+
+```
+cluster://task/default/git-clone
+```
+
+Using the above example, a user can do the following when referencing the Task:
+
+```yaml
+taskRef:
+  resolver: "cluster"
+  name: "cluster://task/default/git-clone" 
+```
+
+### Stage 2
+
+In this stage, we enable the Tekton controller to be able to identify the resolvers on the clusters that have subscribed to the scheme of the incoming [resolver reference](#resolver-reference).
+
+#### Subscribing to `scheme`s
+
+Each remote resolver will subscribe to `scheme`s. e.g. the git resolver might subscribe to `scheme`s like `git+https`, `https` or a bundle resolver might subscribe to `scheme`s like `oci`.
+This mapping of `scheme` to `resolver` could be easily maintained in a `configMap` for the cluster. The cluster operators will be required to know the remote resolvers that are running on the cluster and the schemes they are subscribed to.  
+The configMap must list out all the schemes supported by all the resolvers.
+ 
+```yaml
+apiVersion: v1
+kind: ConfigMap
```

Comment:
I actually am not sure whether we want a configmap for this. we probably need a concept of ordered priority, which we can't easily do with simple key-value pairs (and when parsed go maps can make this ordering non-deterministic).

Reply:
I discuss this in the next section but my idea here was that since the value of the keys is a comma-separated list of resolver names, the order is high to low. So the cluster operators can choose the order of priority based on each scheme.

e.g. say we have a cluster with three resolvers: `resolver-a, resolver-b, resolver-c` and across the resolvers, the supported schemes are: `http, https, oci, git+https`. The config would look like:

```yaml
data:
  https: resolver-a, resolver-b
  http: resolver-b,resolver-a
  oci: resolver-c
  git+https: resolver-a
```

Based on above, config, for the scheme `https`, the cluster operator has decided to provide a higher priority to `resolver-a` over `resolver-b` resolver and vice-versa for the scheme `http`.

Reply:
This approach means the cluster operator needs to install the resolver **and** update some configmap for it to be usable, am I right ?
If that's the case, I am not super fan of the UX (but I am not sure I have something better to propose). In my head, the idea was that resolvers have some default "weight" attached to them (built-in have the highest, so most priority) at build time, and could be overriden / set by operator if they want to change the "default" order, somehow ?

Reply:
Hmm. Good point. 

For built-in resolvers, this is easy and we can order them with some default priority like you suggested.

If users have custom resolvers on the cluster, how would Tekton controller know if that? I thought that a config map like this would help in that case. Even if it can identify other them on the cluster, how does it know about the exact name (what is provided by the `GetName`) of the resolver? In other words, what would the TaskRun/PipelineRun controller set as the label: https://tekton.dev/docs/pipelines/how-to-write-a-resolver/#the-getselector-method?

Reply:
I updated this section and the format of the configMap. PTAL at the design for the updated format. I think with this approach, cluster operators don't need to do anything. When the framework calls the `Initialize` method, the resolvers register their schemes to the configMap that Tekton reads from.

Reply:
We definitely need some sort of data source to configure these - I just don't know if a ConfigMap is the right tool. 🤔 

If multiple resolvers subscribe to a single scheme type it's ambiguous what order that will happen in since ConfigMap key ordering is not guaranteed to remain stable after parsing due to the nature of Go maps. So we'd either need to write our own ConfigMap parser, use a different structure, or use some clever string encoding (though idk if this will be the most readable).

The alternative format you put in https://github.com/tektoncd/community/pull/1138#discussion_r1536674203 is less ambiguous since resolver ordering will be retained in the value. We should firm up the order of operations in particular for `*` (my guess is this should be a catch-all and evaluated last). The trade-off here is that it's more of a pain to add a new resolver - you can't add a new resolver simply by patching in a line addition.

I don't think we should grant configmap rbac to resolvers, though I like that you're thinking about the UX around managing resolvers!

Reply:
I've updated the design based on our discussions. PTAL again.